crates/walrus-core/src/test_utils.rs

```diff
@@ -38,6 +47,135 @@ use crate::{
+    /// A map of blob identifiers, keyed by tag keys and then tag values.
+    pub blob_identifiers_by_tag: HashMap<String, HashMap<String, HashSet<String>>>,
+}
+
+impl<'a> QuiltTestData<'a> {
+    /// Generates a new QuiltTestData from existing data.
+    pub fn new(
+        blob_data: &'a [&'a [u8]],
+        max_value_length: usize,
+        max_num_tags: usize,
+    ) -> Result<Self, QuiltError> {
+        let cow_data: Vec<Cow<'a, [u8]>> =
+            blob_data.iter().map(|&data| Cow::Borrowed(data)).collect();
+
+        let quilt_store_blobs_vec =
+            Self::generate_random_quilt_store_blobs(cow_data, max_value_length, max_num_tags)?;
+
+        Self::from_quilt_store_blobs(quilt_store_blobs_vec)
+    }
+
+    /// Generates a new QuiltTestData with new owned random data.
+    pub fn new_owned(
+        num_blobs: usize,
+        min_blob_size: usize,
+        max_blob_size: usize,
+        max_value_length: usize,
+        max_num_tags: usize,
+    ) -> Result<QuiltTestData<'static>, QuiltError> {
+        let blob_data =
+            walrus_test_utils::generate_random_data(num_blobs, min_blob_size, max_blob_size);
+
+        let cow_data: Vec<Cow<'static, [u8]>> = blob_data.into_iter().map(Cow::Owned).collect();
+
+        let quilt_store_blobs_vec =
+            Self::generate_random_quilt_store_blobs(cow_data, max_value_length, max_num_tags)?;
+
+        QuiltTestData::from_quilt_store_blobs(quilt_store_blobs_vec)
+    }
+
+    /// Returns the quilt store blobs and clears the internal vector.
+    pub fn take_blobs(&mut self) -> Vec<QuiltStoreBlob<'a>> {
+        core::mem::take(&mut self.quilt_store_blobs)
+    }
+
+    fn from_quilt_store_blobs(
+        quilt_store_blobs: Vec<QuiltStoreBlob<'a>>,
+    ) -> Result<Self, QuiltError> {
+        let mut blob_identifiers_by_tag: HashMap<String, HashMap<String, HashSet<String>>> =
+            HashMap::new();
+
+        for blob in &quilt_store_blobs {
+            for (key, value) in blob.tags() {
+                blob_identifiers_by_tag
+                    .entry(key.clone())
+                    .or_default()
+                    .entry(value.clone())
+                    .or_default()
+                    .insert(blob.identifier().to_string());
+            }
+        }
+
+        Ok(Self {
+            quilt_store_blobs,
+            blob_identifiers_by_tag,
+        })
+    }
+
+    /// Generates random QuiltStoreBlobs from the input raw blobs.
+    ///
+    /// A random unique identifier is generated for each blob.
+    /// Random numbers of random tags are generated for each blob.
+    fn generate_random_quilt_store_blobs<'b>(
+        blob_data: Vec<Cow<'b, [u8]>>,
+        max_value_length: usize,
+        max_num_tags: usize,
+    ) -> Result<Vec<QuiltStoreBlob<'b>>, QuiltError> {
+        tracing::debug!("generating random quilt store blobs...");
+        let mut rng = rand::thread_rng();
+        let num_tags = rng.gen_range(0..=max_num_tags);
```

Comment:
Nit: I wonder if num_tags = 0 is more interesting and should have higher probability than other values.

Also, do we allow 0 byte blob inside a quilt?